righttyper/unified_transformer.py

```diff
@@ -421,21 +439,53 @@ def _process_parameter(self, parameter: cst.Param, ann: FuncAnnotation) -> cst.P
 
     def leave_FunctionDef(
             self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef
-    ) -> cst.FunctionDef | cst.FlattenSentinel:
+    ) -> cst.FunctionDef | cst.FlattenSentinel | cst.RemovalSentinel:
         name = ".".join(self.name_stack[:-1])
         self.name_stack.pop()
         self.name_stack.pop()
         self.used_names.pop()
+        self.overload_stack.pop()
+        self.overload_name_stack.pop()
 
         first_line = min(
             self.get_metadata(PositionProvider, node).start.line
             for node in (original_node, *original_node.decorators)
         )
         key = FuncId(Filename(self.filename), first_line, FunctionName(name))
 
+        # If we encounter a function whose name doesn't match the current
+        # overload sequence, we discard the overload stack and update the
+        # current overload sequence.
+        # This also has the side effect of cleaning up orphan overload
+        # signatures, which is good.
+        if name != self.overload_name_stack[-1]:
+            self.overload_stack[-1] = []
+            self.overload_name_stack[-1] = name
+
+        is_overload = any(
+            decorator.decorator.value == "overload"
+            for decorator in updated_node.decorators
+            if isinstance(decorator.decorator, cst.Name)
+        )
+        # We also look for qualified references to typing.overload
+        is_overload = is_overload | any(
+            decorator.decorator.attr.value == "overload"
+            for decorator in updated_node.decorators
+            if isinstance(decorator.decorator, cst.Attribute)
```

Comment:
This only checks that the decorator is either named "overload" or has an attribute named "overload".
I don't feel we need to support every possible way `typing.overload` can be aliased, but we need to at least support however `overload` is being imported, kind of how we keep track of `self.aliases` (but not quite -- that serves another purpose).